File: 그래프 부수기/BOJ_16929.java
package BOJ_16929;

import java.io.*;
import java.util.StringTokenizer;

public class BOJ_16929 {

    // https://www.acmicpc.net/problem/16929
    // input
    private static BufferedReader br;

    // variables
    private static int N, M;
    private static char[][] board;
    private static boolean[][] isVisited;
    private static final int[] dirX = {-1, 0, 1, 0}; // 상 우 하 좌
    private static final int[] dirY = {0, 1, 0, -1};

    public static void main(String[] args) throws IOException {
        System.setIn(new FileInputStream("C:\\Users\\bigyo\\Desktop\\알고리즘\\JavaAlgorithm\\src\\BOJ_16929\\res.txt"));
        br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        input();

        bw.write(solve());
        bw.close();
    } // End of main()

    private static String solve() {
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (isVisited[i][j]) continue;

                if (DFS(i, j, i, j, board[i][j])) {
                    sb.append("Yes");
                    return sb.toString();
                }
            }
        }


        sb.append("No");
        return sb.toString();
    } // End of solve()

    private static boolean DFS(int x, int y, int preX, int preY, char color) {
        if (isVisited[x][y]) return true;
        isVisited[x][y] = true;

        for (int i = 0; i < 4; i++) {
            int nX = dirX[i] + x;
            int nY = dirY[i] + y;

            if (!isAbleCheck(nX, nY, color)) continue;
            if (nX != preX || nY != preY) {
                if (DFS(nX, nY, x, y, color)) return true;
            }
        }

        return false;
    } // End of DFS()

    private static boolean isAbleCheck(int x, int y, char color) {
        return x >= 0 && x < N && y >= 0 && y < M && board[x][y] == color;
    } // End of isAbleCheck()

    private static void input() throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        board = new char[N][M];
        isVisited = new boolean[N][M];

        for (int i = 0; i < N; i++) {
            board[i] = br.readLine().toCharArray();
        }
    } // End of input()
} // End of Main class

File: 그래프 부수기/BOJ 16929.kt
package BOJ_16929

import java.io.File
import java.util.StringTokenizer

// https://www.acmicpc.net/problem/16929
// input
private var br = System.`in`.bufferedReader()

// variables
private var N = 0
private var M = 0
private lateinit var arr: Array<CharArray>
private lateinit var isVisited: Array<BooleanArray>

private val dirX = intArrayOf(-1, 0, 1, 0)
private val dirY = intArrayOf(0, 1, 0, -1)

fun main() {
    val path = "C:\\Users\\bigyo\\Desktop\\알고리즘\\KotlinAlgo\\src\\main\\kotlin\\BOJ_16929\\res.txt"
    br = File(path).bufferedReader()
    val bw = System.out.bufferedWriter()

    input()

    bw.write(solve())
    bw.close()
} // End of main()

private fun solve(): String {
    val sb = StringBuilder()

    for (i in 0 until N) {
        for (j in 0 until M) {
            if (isVisited[i][j]) continue

            if (DFS(i, j, i, j, arr[i][j])) {
                sb.append("Yes")
                return sb.toString()
            }
        }
    }

    sb.append("No")
    return sb.toString()
} // End of solve()

private fun DFS(x: Int, y: Int, preX: Int, preY: Int, color: Char): Boolean {
    if (isVisited[x][y]) return true
    isVisited[x][y] = true

    for (i in 0 until 4) {
        val nX = x + dirX[i]
        val nY = y + dirY[i]

        if (!isAbleCheck(nX, nY, color)) continue
        if (nX == preX && nY == preY) continue

        if (DFS(nX, nY, x, y, color)) {
            return true
        }
    }

    return false
} // End of DFS()

private fun isAbleCheck(x: Int, y: Int, color: Char): Boolean {
    return x in 0 until N && y in 0 until M && arr[x][y] == color
} // End of isAbleCheck()

private fun input() {
    val st = StringTokenizer(br.readLine())
    N = st.nextToken().toInt()
    M = st.nextToken().toInt()

    arr = Array(N) {
        br.readLine().toCharArray()
    }
    isVisited = Array(N) { BooleanArray(M) }
} // End of input()
